mem-ruby: Move GPU L1 cache MSHR to the coalescer

v-ramadas opened the pull request:
This commit undoes PR https://github.com/gem5/gem5/pull/540 which added MSHRs to the L1 SQC/TCP to handle requests to the same line. The commit moves the MSHR instead to the coalescer where the coalescedTable maintains a list of all requests that go to the same line. These pending requests are completed as soon as the first request that went to cache returns. If any of these pending requests have GLC/SLC flags set, then the request is not marked complete. It is instead sent to cache. The commit also adds stats to track coalescer mshr misses and hits under misses

mattsinc commented on configs/example/gpufs/runfs.py:
```diff
@@ -291,8 +291,12 @@ def runGpuFSSystem(args):
                 break
             kernels_completed += 1
             tasks_completed += 1
+            m5.stats.dump()
```
I don't think these should be part of this commit

v-ramadas replied:
Oops. I think they got added by mistake; removed now. The PR I created earlier for the stats dump and reset got rejected and I guess the commit somehow got stuck to this PR...

mattsinc commented on src/mem/ruby/protocol/GPU_VIPER-SQC.sm:
```diff
@@ -48,9 +48,6 @@ machine(MachineType:SQC, "GPU SQC (L1 I Cache)")
 {
   state_declaration(State, desc="SQC Cache States", default="SQC_State_I") {
     I, AccessPermission:Invalid, desc="Invalid";
-    // Note: currently IV in the TCP is only for pending loads to a given cache
```
Does all of this code need to be removed to change how the MSHRs work?  I guess your point is the states are no longer reachable?

v-ramadas replied:
I removed these lines because the IV state is no longer reachable.

mattsinc commented on src/mem/ruby/protocol/GPU_VIPER-TCP.sm:
```diff
@@ -396,7 +392,7 @@ machine(MachineType:TCP, "GPU TCP (L1 Data Cache)")
         sequencer.writeCallback(address, tmp, false, MachineType:L1Cache);
       } else {
-        coalescer.writeCallback(address, MachineType:L1Cache, tmp);
+        coalescer.writeCallback(address, MachineType:L1Cache, tmp, false);
```
is this change part of the MSHR fix?

v-ramadas replied:
It's part of the MSHR stats fix. Will make that a separate commit tonight

mattsinc replied:
ok, let me know when these are separate.

v-ramadas replied:
Sorry, I got sidetracked yesterday. Just updated the PR after making these separate

mattsinc commented on src/mem/ruby/system/GPUCoalescer.cc:
```diff
@@ -211,6 +211,7 @@ GPUCoalescer::GPUCoalescer(const Params &p)
                  false, Event::Progress_Event_Pri),
       uncoalescedTable(this),
       deadlockCheckEvent([this]{ wakeup(); }, "GPUCoalescer deadlock check"),
+      stats(this),
```
Stats should be a separate commit?

v-ramadas replied:
Will make it a separate commit

v-ramadas committed:
mem-ruby: Move GPU L1 cache MSHR to the coalescer

This commit undoes PR #540 which added MSHRs to the L1 SQC/TCP to handle
requests to the same line. The commit moves the MSHR instead to the
coalescer where the coalescedTable maintains a list of all requests that
go to the same line. These pending requests are completed as soon as the
first request that went to cache returns. If any of these pending
requests have GLC/SLC flags set, then the request is not marked
complete. It is instead sent to cache.

v-ramadas committed:
mem-ruby: Add GPUCoalescer MSHR stats

This commit adds stats to the GPUCoalescer to track mshr misses and
hits-under-misses. The TCP hit counter is also incremented for mshr
hits-under-misses as MSHRs at the coalescer is equivalent to using TBEs
in the TCP Ruby protocol. Since coalescer MSHR misses go down to the
TCP, the miss counters are not incremented here

abmerop commented on src/mem/ruby/system/GPUCoalescer.cc:
```diff
@@ -1025,12 +1050,52 @@ GPUCoalescer::completeHitCallback(std::vector<PacketPtr> & mylist)
 void
-GPUCoalescer::recordMissLatency(CoalescedRequest* crequest,
+GPUCoalescer::recordStats(CoalescedRequest* crequest,
                                 MachineType mach,
                                 Cycles initialRequestTime,
                                 Cycles forwardRequestTime,
                                 Cycles firstResponseTime,
-                                bool success, bool isRegion)
+                                bool isRegion, bool mshrHitUnderMiss)
```
Nit: Are all these parameters needed?

v-ramadas replied:
isRegion isn't used but was there in the older version. Same with the Cycles parameters; they were used to calculate a few stats long ago before the current stats class was introduced. But they can't be used anymore since some readCallback() calls set them to zero. I didn't remove them assuming they might be used in the future. Do you think its better to remove them now, and add them again if needed later?

abmerop replied:
Probably fine to keep until we know if they are used in other protocols but I assume they are not (publicly).

abmerop commented on src/mem/ruby/system/GPUCoalescer.cc:
```diff
@@ -530,16 +531,38 @@ GPUCoalescer::readCallback(Addr address,
+    // Coalescer. Do not respond to pending loads that have SLC/GLC flags
+    // set; issue them instead
+    while (crequest->getRubyType() == RubyRequestType_LD) {
+    hitCallback(crequest, mach, data, true,
+            crequest->getIssueTime(), forwardRequestTime, firstResponseTime,
+            isRegion, externalHit, mshr_hit_under_miss);
+
```
Nit: Alignment

abmerop reviewed (commented):
I believe we did something similar in the past. The issue was the hit/miss stats printed by Ruby were not accurate. Is the intention to combine the new stats here + Ruby stats or are you planning a way to update the Ruby stats too?

mattsinc commented:
> I believe we did something similar in the past. The issue was the hit/miss stats printed by Ruby were not accurate. Is the intention to combine the new stats here + Ruby stats or are you planning a way to update the Ruby stats too?

@v-ramadas : I believe the fixes between this and the stats here, you have validated they give correct results now?

mattsinc reviewed (commented):
This looks ok to me now, but will wait for @abmerop's comments to be addresed

v-ramadas commented:
@abmerop, @mattsinc, the new stats I added track the MSHR's effectiveness. They also update the TCP hit rate when there is an MSHR hit-under-miss. I validated them for a small test that does three loads to the same address; I see that one MSHR miss and two MSHR hits-under-misses. I also see that the TCP cache hits are updated with the two MSHR hits under misses. Additionally, if the GLC or SLC bit is set, requests go to the TCP and do not hit in the MSHR

mattsinc reviewed (approved):
Ok on my end, but will wait for @abmerop before pushing

mattsinc commented:
> Ok on my end, but will wait for @abmerop before pushing

@abmerop just checking on this
